File: Binary Tree Level Order Traversal II/Binary_Tree_Level_Order_Traversal_II.py
# 64ms 36.69%

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        values = []
        if root is None:
            return values
        
        queue = []
        queue.append(root)

        while len(queue) > 0:
            value = []
            size = len(queue)
            while size > 0:
                tempNode = queue.pop(0)
                value.append(tempNode.val)

                if tempNode.left:
                    queue.append(tempNode.left)
                if tempNode.right:
                    queue.append(tempNode.right)

                size -= 1
            values.append(value)

        return values[::-1]

File: Binary Tree Level Order Traversal II/Binary_Tree_Level_Order_Traversal_II.cpp
// 和102思路一致 区别在于使用堆栈来存储
// 4ms 99.02%

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include<vector>
#include<queue>
#include<stack>
using namespace std;

class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        
        vector<vector<int>> values;
        if (!root)return values;

        stack<vector<int>> valueStack;

        queue<TreeNode*> treeQueue;
        treeQueue.push(root);
        
        while (!treeQueue.empty())
        {
            vector<int> tempVec;
            int size = treeQueue.size();

            while (size--)
            {
                TreeNode *tempNode = treeQueue.front();
                treeQueue.pop();

                tempVec.push_back(tempNode->val);

                if (tempNode->left)
                    treeQueue.push(tempNode->left);
                if (tempNode->right)
                    treeQueue.push(tempNode->right);
                
            }
            valueStack.push(tempVec);
        }

        while (!valueStack.empty())
        {
            values.push_back(valueStack.top());
            valueStack.pop();
        }

        // 或者使用下边的一句话结束
        // reverse(values.begin(), values.end());
        return values;
    }
};